### PR TITLE
Use x509 Authentication endpoint URL when configured

### LINKED ISSUE #17484: x509 certificate authenticator configured AuthenticationEndpoint URL is not honored when redirecting to commonauth and authorize endpoints
**Describe the issue:**
When configuring x509 Authenticator the following configurations are included in deployment.toml file to configure AuthenticationEndpoint URL

[authentication.authenticator.x509_certificate.parameters]
name ="x509CertificateAuthenticator"
enable=true
AuthenticationEndpoint="[https://sswso2is.local:443/x509-certificate-servlet](https://sswso2is.local/x509-certificate-servlet)"
username= "CN"

But after the successful response from the configured AuthenticationEndpoint url above ([https://sswso2is.local:443/x509-certificate-servlet?<params](https://sswso2is.local/x509-certificate-servlet?%3Cparams)>), user redirects to https://<Identity_server_hostname>:8443/commonauth?sessionDataKey=<>&success=true instead of redirecting to configured Authenticationendpoint URL domain (ex: [https://sswso2is.local:443/commonauth](https://sswso2is.local/commonauth)). This happens because of [1]. Same happens when redirecting to /oauth2/authorize when redirecting from above mentioned commonauth endpoint.

**How to reproduce:**
This can be reproduced using following steps

1. Set up x509 authenticator as second step using [2]
2. Set up fronting nginx
3. configure different domain name from Identity server hostname for AuthenticationEndpoint using following configuration deployment.toml

[authentication.authenticator.x509_certificate.parameters]
name ="x509CertificateAuthenticator"
enable=true
AuthenticationEndpoint="[https://sswso2is.local:443/x509-certificate-servlet](https://sswso2is.local/x509-certificate-servlet)"
username= "CN"

4. Then map configured domain name in AuthenticationEndpoint URL config to wso2 identity server’s 8443 port in nginx
5. Configure x509 CertificateAuthenticator as second step in authentication flow

**Expected behavior:**
After successful response from configured x509-certificate-servlet endpoint (AuthenticationEndpoint="https://sswso2is.local:443/x509-certificate-servlet"), redirect user to the commonauth endpoint of configured domain name in above AuthenticationEndpoint URL (ex: https://sswso2is.local:443/commonauth) and then to oauth2/authorize endpoint instead of redirecting to https://<Identity_server_hostname>:8443/commonauth and https://<Identity_server_hostname>:8443/oauth2/authorize.

**Environment information** (_Please complete the following information; remove  any unnecessary fields_) **:**
 - Product Version: IS 5.11 
 - OS: [e.g., Windows, Linux, Mac]
 - Database: [e.g., MySQL, H2]
 - Userstore: [e.g., LDAP, JDBC]

---
[1] - https://github.com/wso2-extensions/identity-outbound-auth-x509/blob/82c502d07e0b4247ed442b6cab57e143d16d0e18/component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java#L62

[2] - https://is.docs.wso2.com/en/5.11.0/learn/x509certificate-authenticator/
### Optional Fields

**Related issues:**

**Suggested labels:**

### PR DESCRIPTION
**Pull request by mpmadhavig**
Resolves https://github.com/wso2/product-is/issues/17484

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java [hunk 1]
```diff
@@ -69,6 +71,21 @@ protected void doPost(HttpServletRequest servletRequest, HttpServletResponse ser
...
+
+        if (authenticationEndpoint != null) {
+            URL authenticationEndpointURL = new URL(authenticationEndpoint);
+            String protocol = authenticationEndpointURL.getProtocol();
+            String host = authenticationEndpointURL.getHost();
+            int port = authenticationEndpointURL.getPort();
+            if (port == -1) {
+                commonAuthURL = protocol + "://" + host + "/" + X509CertificateConstants.COMMON_AUTH;
+            } else {
+                commonAuthURL = protocol + "://" + host + ":" + port + "/" + X509CertificateConstants.COMMON_AUTH;
```
Comment (AmshikaH) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java]: Can we move `://`, `:` and `/` to constants?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java [hunk 2]
```diff
@@ -69,6 +71,21 @@ protected void doPost(HttpServletRequest servletRequest, HttpServletResponse ser
             throw new RuntimeException("Error occurred while building URL.", e);
         }
 
+        String authenticationEndpoint = X509CertificateUtil.getX509Parameters().get(X509CertificateConstants
+                .AUTHENTICATION_ENDPOINT_PARAMETER);
+
+        if (authenticationEndpoint != null) {
```
Comment (ShanChathusanda93) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java]: Shall we use the StringUtils?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java (hunk 1 above)
Comment (ShanChathusanda93) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java]: AFAIU the commonAuthURL will not contain the port number if the port is only `-1`. Can this port be a different minus value like `-2`, `-3` etc. If so is it ok to add that to the URL?

Comment (mpmadhavig) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java]: Resolved with 19a2fddeacdc3ffdeae1d0ad4429a6be8bb576de

Comment (mpmadhavig) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java]: According to the documentation of java.net.URL getPort[1] method, `Returns: the port number, or -1 if the port is not set`.

[1] https://docs.oracle.com/javase/8/docs/api/java/net/URL.html#getPort--

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java (hunk 2 above)
Comment (mpmadhavig) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateServlet.java]: Resolved with 875f79fbf7010e5b154a018f76df1c044f26184d

**Pull request closed without merging**